sdks/python/apache_beam/yaml/yaml_ml.py

```diff
@@ -33,11 +40,419 @@
+    config = spec.get('config')
+    for required in ('model_handler', ):
+      if required not in config:
+        raise ValueError(
+            f'Missing {required} parameter in RunInference config '
+            f'at line {SafeLineLoader.get_line(spec)}')
+    model_handler = config.get('model_handler')
+    if not isinstance(model_handler, dict):
+      raise ValueError(
+          'Invalid model_handler specification at line '
+          f'{SafeLineLoader.get_line(spec)}. Expected '
+          f'dict but was {type(model_handler)}.')
+    for required in ('type', 'config'):
+      if required not in model_handler:
+        raise ValueError(
+            f'Missing {required} in model handler '
+            f'at line {SafeLineLoader.get_line(model_handler)}')
+    typ = model_handler['type']
+    extra_params = set(SafeLineLoader.strip_metadata(model_handler).keys()) - {
+        'type', 'config'
+    }
+    if extra_params:
+      raise ValueError(
+          f'Unexpected parameters in model handler of type {typ} '
+          f'at line {SafeLineLoader.get_line(spec)}: {extra_params}')
+    model_handler_provider = ModelHandlerProvider.handler_types.get(typ, None)
+    if model_handler_provider:
+      model_handler_provider.validate(model_handler['config'])
+    else:
+      raise NotImplementedError(
+          f'Unknown model handler type: {typ} '
+          f'at line {SafeLineLoader.get_line(spec)}.')
+
+  return spec
+
+
+class ModelHandlerProvider:
+  handler_types: Dict[str, Callable[..., "ModelHandlerProvider"]] = {}
+
+  def __init__(
+      self, handler, preprocess: Callable = None, postprocess: Callable = None):
+    self._handler = handler
+    self._preprocess = self.parse_processing_transform(
+        preprocess, 'preprocess') or self.preprocess_fn
+    self._postprocess = self.parse_processing_transform(
+        postprocess, 'postprocess') or self.postprocess_fn
+
+  def get_output_schema(self):
+    return Any
+
+  @staticmethod
+  def parse_processing_transform(processing_transform, typ):
+    def _parse_config(callable=None, path=None, name=None):
+      if callable and (path or name):
+        raise ValueError(
+            f"Cannot specify 'callable' with 'path' and 'name' for {typ} "
+            f"function.")
+      if path and name:
+        return python_callable.PythonCallableWithSource.load_from_script(
+            FileSystems.open(path).read().decode(), name)
+      elif callable:
+        return python_callable.PythonCallableWithSource(callable)
+      else:
+        raise ValueError(
+            f"Must specify one of 'callable' or 'path' and 'name' for {typ} "
+            f"function.")
+
+    if processing_transform:
+      if isinstance(processing_transform, dict):
+        return _parse_config(**processing_transform)
+      else:
+        raise ValueError("Invalid model_handler specification.")
+
+  def underlying_handler(self):
+    return self._handler
+
+  def preprocess_fn(self, row):
```

Comment:
Great suggestion, I ended up making the preprocess parameter required on the VertexAI, so the error is redundant in this case... but I think moving forward it could be useful, and the refactor in general is cleaner.